services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/AsyncMappedIndex.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.extensions.dynamodb.mappingclient;
+
+import java.util.concurrent.CompletableFuture;
+
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.async.SdkPublisher;
+
+/**
+ * Asynchronous interface for running commands against an object that is linked to a specific DynamoDb secondary index
+ * and knows how to map records from the table that index is linked to into a modelled object.
+ *
+ * Typically an implementation for this interface can be obtained from a {@link AsyncMappedTable} which in turn can be
+ * obtained from a {@link AsyncMappedDatabase}:
+ *
+ * mappedIndex = mappedDatabase.table(tableSchema).index("gsi_1");
```

Comment:
I'm going to remove the code examples, because it's too difficult to keep them up to date as we iterate on usability. Maybe we will add them back in once we've locked the interface.